src/cursor/cur_hs.c

```diff
@@ -185,6 +184,151 @@ __curhs_reset(WT_CURSOR *cursor)
+    WT_DECL_RET;
+    va_list ap;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+
+    va_start(ap, cursor);
+    ret = file_cursor->get_key(file_cursor, va_arg(ap, uint32_t *), va_arg(ap, WT_ITEM **),
+      va_arg(ap, wt_timestamp_t *), va_arg(ap, uint64_t *));
+    va_end(ap);
+
+    return (ret);
+}
+
+/*
+ * __curhs_get_value --
+ *     WT_CURSOR->get_value method for the hs cursor type.
+ */
+static int
+__curhs_get_value(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_RET;
+    va_list ap;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+
+    va_start(ap, cursor);
+    ret = file_cursor->get_value(file_cursor, va_arg(ap, wt_timestamp_t *),
+      va_arg(ap, wt_timestamp_t *), va_arg(ap, uint64_t *), va_arg(ap, WT_ITEM **));
+    va_end(ap);
+
+    return (ret);
+}
+
+/*
+ * __curhs_set_value --
+ *     WT_CURSOR->set_value method for the hs cursor type.
+ */
+static void
+__curhs_set_value(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_HS *hs_cursor;
+    va_list ap;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+    va_start(ap, cursor);
+    hs_cursor->time_window = *va_arg(ap, WT_TIME_WINDOW *);
+
+    file_cursor->set_value(file_cursor, va_arg(ap, wt_timestamp_t), va_arg(ap, wt_timestamp_t),
+      va_arg(ap, uint64_t), va_arg(ap, WT_ITEM *));
+    va_end(ap);
+}
+
+/*
+ * __curhs_insert --
+ *     WT_CURSOR->insert method for the hs cursor type.
+ */
+static int
+__curhs_insert(WT_CURSOR *cursor)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *hs_upd, *upd_local;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+    cbt = (WT_CURSOR_BTREE *)file_cursor;
+    hs_upd = upd_local = NULL;
+
+    CURSOR_API_CALL_PREPARE_ALLOWED(cursor, session, insert, CUR2BT(file_cursor));
+
+    /* Allocate a tombstone only when there is a valid stop time point. */
+    if (WT_TIME_WINDOW_HAS_STOP(&hs_cursor->time_window)) {
+        /*
+         * Insert a delete record to represent stop time point for the actual record to be inserted.
+         * Set the stop time point as the commit time point of the history store delete record.
+         */
+        WT_ERR(__wt_upd_alloc_tombstone(session, &hs_upd, NULL));
+        hs_upd->start_ts = hs_cursor->time_window.stop_ts;
+        hs_upd->durable_ts = hs_cursor->time_window.durable_stop_ts;
+        hs_upd->txnid = hs_cursor->time_window.stop_txn;
+    }
+
+    /*
+     * Append to the delete record, the actual record to be inserted into the history store. Set the
+     * current update start time point as the commit time point to the history store record.
+     */
+    WT_ERR(__wt_upd_alloc(session, &file_cursor->value, WT_UPDATE_STANDARD, &upd_local, NULL));
```

Comment:
Don't we need to free this memory in the error handling of the function?

Reply:
fixed.